### PR TITLE
Introduce simple remote connection strategy

### PR DESCRIPTION
**Pull request by Tim-Brooks**
This commit introduces a simple remote connection strategy which will
open remote connections to a configurable list of user supplied
addresses. These addresses can be remote Elasticsearch nodes or
intermediate proxies. We will perform normal clustername and version
validation, but otherwise rely on the remote cluster to route requests
to the appropriate remote node.

### CODE REVIEW

File: server/src/test/java/org/elasticsearch/transport/SimpleConnectionStrategyTests.java [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+                         numOfConnections, addresses(address1, address2))) {
+                    assertFalse(connectionManager.getAllConnectedNodes().stream().anyMatch(n -> n.getAddress().equals(address1)));
+                    assertFalse(connectionManager.getAllConnectedNodes().stream().anyMatch(n -> n.getAddress().equals(address2)));
+
+                    PlainActionFuture<Void> connectFuture = PlainActionFuture.newFuture();
+                    strategy.connect(connectFuture);
+                    connectFuture.actionGet();
+
+                    assertTrue(connectionManager.getAllConnectedNodes().stream().anyMatch(n -> n.getAddress().equals(address1)));
+                    assertFalse(connectionManager.getAllConnectedNodes().stream().anyMatch(n -> n.getAddress().equals(address2)));
```
Comment (ywelsch) [inline server/src/test/java/org/elasticsearch/transport/SimpleConnectionStrategyTests.java]: Can we check how many connections have been established?

File: server/src/test/java/org/elasticsearch/transport/SimpleConnectionStrategyTests.java [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+        }
+    }
+
+    private static List<Supplier<TransportAddress>> addresses(final TransportAddress... addresses) {
+        if (addresses.length == 0) {
+            return Collections.emptyList();
+        } else if (addresses.length == 1) {
+            return Collections.singletonList(() -> addresses[0]);
+        } else {
+            return Arrays.stream(addresses)
```
Comment (ywelsch) [inline server/src/test/java/org/elasticsearch/transport/SimpleConnectionStrategyTests.java]: why not always return this? Doesn't this work for the `length = 0` and `length = 1` cases?

File: server/src/test/java/org/elasticsearch/transport/SimpleConnectionStrategyTests.java (hunk 1 above)
Comment (ywelsch) [reply server/src/test/java/org/elasticsearch/transport/SimpleConnectionStrategyTests.java]: Can we check the number of connection attempts?

**Review by ywelsch — CHANGES_REQUESTED** [review]
Thanks for the PR. I've left some suggestions for code simplification

**Comment by tbrooks8** [comment]
@ywelsch i updated this PR for your comments.

File: server/src/main/java/org/elasticsearch/transport/RemoteConnectionStrategy.java [hunk 3]
```diff
@@ -161,4 +166,19 @@ boolean assertNoRunningConnections() {
         }
         return result;
     }
+
+    static Predicate<ClusterName> getRemoteClusterNamePredicate(ClusterName remoteClusterName) {
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/transport/RemoteConnectionStrategy.java]: this is only used by SniffConnectionStrategy now. Move it back down?

File: server/src/main/java/org/elasticsearch/transport/SimpleConnectionStrategy.java [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+    }
+
+    @Override
+    protected void connectImpl(ActionListener<Void> listener) {
+        performSimpleConnectionProcess(addresses.iterator(), listener);
+    }
+
+    private void performSimpleConnectionProcess(Iterator<Supplier<TransportAddress>> addressIter, ActionListener<Void> listener) {
+        openConnections(listener, 1);
+
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/transport/SimpleConnectionStrategy.java]: extra newline

File: server/src/main/java/org/elasticsearch/transport/SimpleConnectionStrategy.java [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+                    public void onFailure(Exception e) {
+                        logger.debug(() -> new ParameterizedMessage("failed to open remote connection to address {}", address), e);
+                        compositeListener.onFailure(e);
+                    }
+                });
+            }
+        } else {
+            int openConnections = connectionManager.size();
+            if (openConnections == 0) {
+                finished.onFailure(new IllegalStateException("Unable to open any simple connections to remote cluster"));
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/transport/SimpleConnectionStrategy.java]: mention remote cluster name here?

File: server/src/main/java/org/elasticsearch/transport/SimpleConnectionStrategy.java [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+            } else {
+                logger.trace("unable to open maximum number of connections [opened: {}, maximum: {}]", openConnections,
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/transport/SimpleConnectionStrategy.java]: mention remote cluster name here, and log at debug level?

trace is for expected things, debug for more unusual stuff

File: server/src/main/java/org/elasticsearch/transport/SimpleConnectionStrategy.java [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+                ConnectionManager.ConnectionValidator validator = clusterNameValidator(node);
+                connectionManager.connectToNode(node, profile, validator, new ActionListener<>() {
+                    @Override
+                    public void onResponse(Void v) {
+                        compositeListener.onResponse(v);
+                    }
+
+                    @Override
+                    public void onFailure(Exception e) {
+                        logger.debug(() -> new ParameterizedMessage("failed to open remote connection to address {}", address), e);
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/transport/SimpleConnectionStrategy.java]: mention cluster name here

File: server/src/main/java/org/elasticsearch/transport/SimpleConnectionStrategy.java [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+        }
+    }
+
+    private TransportAddress nextAddress(List<TransportAddress> resolvedAddresses) {
+        long curr;
+        while ((curr = counter.getAndIncrement()) == Long.MIN_VALUE) ;
+        return resolvedAddresses.get(Math.floorMod(curr, resolvedAddresses.size()));
+    }
+
+    private ConnectionManager.ConnectionValidator clusterNameValidator(DiscoveryNode node) {
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/transport/SimpleConnectionStrategy.java]: let's remove the `node` parameter here and have a single instance of the `ConnectionValidator` initialized in the constructor. The node here is not that useful, and you can get it from the connection (`newConnection.getNode()`).

**Review by ywelsch — APPROVED** [review]
I've left some nits, otherwise looking good.

**Comment by fpytloun** [comment]
This is great feature, thank you.
Can you please post example configuration to switch existing setup to using simple strategy?

**Comment by tbrooks8** [comment]
@fpytloun The meta issue is here #49067. As you can see, the work is scoped for 7.6 although things can always change.

Currently the "simple" work is currently available on either 7.6 or 8.0 snapshot builds.

It can be configured:

```
// If you do not configure a mode, it will default to the existing sniff implementation
cluster.remote.test_remote_cluster.mode: "simple"
cluster.remote.test_remote_cluster.simple.addresses: ["address:port"]
```

Please note that it is still in active development. For example, we are still debating whether we want to name it "simple" or "proxy" (the primary advantage of this mode over sniff is for routing through a proxy to unaddressable nodes).

You can follow any updates on the meta issue and the documentation will be updated before any release where this feature is GA.

**Pull request merged**